New package: hashcash-1.23

g4s8 opened the pull request:
<!-- Mark items with [x] where applicable -->
Added new `hashcash` package for www.hashcash.org/ 
It's currently available in default Debian and Ubuntu repositories, but not in Void repo.

#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

<!--
If GitHub CI cannot be used to validate the build result (for example, if the
build is likely to take several hours), make sure to
[skip CI](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration).
When skipping CI, uncomment and fill out the following section.
Note: for builds that are likely to complete in less than 2 hours, it is not
acceptable to skip CI.
-->

#### Does it build and run successfully? 
(Please choose at least one native build and, if supported, at least one cross build. More are better.)
- [x] I built this PR locally for my native architecture, (ARCH-LIBC)
- [ ] I built this PR locally for these architectures (if supported. mark crossbuilds):
  - [ ] aarch64-musl
  - [ ] armv7l
  - [ ] armv6l-musl

ericonr commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+homepage="http://www.hashcash.org"
+distfiles="http://www.hashcash.org/source/hashcash-${version}-pr.tgz"
+checksum=03a85b6a8a84be93f76a7297fccd5f61dcb8255de25f3fbd77c26be3ef9be999
+make_build_args+=" hashcash docs"
+make_install_args+=" install"
```
You should just set these, they don't have an initial value. Also, put this directly below `build_style`.

ericonr commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+revision=1
+archs="~*-musl"
+build_style=gnu-makefile
+hostmakedepends="make"
```
This is unnecessary *and* breaks the masterdir.

ericonr commented:
Also, this is bad code

```
array.c:13:33: warning: implicit declaration of function 'die_msg' [-Wimplicit-function-declaration]
   13 |     if ( array->elt == NULL ) { die_msg( "out of memory" ); }
      |                                 ^~~~~~~
sdb.c: In function 'sdb_cb_token_matcher':
sdb.c:319:3: warning: implicit declaration of function 'die_msg' [-Wimplicit-function-declaration]
  319 |   die_msg( re_err );
      |   ^~~~~~~
```

And conflicts with `outils`

Finally, what's it useful for standalone?

g4s8 commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+pkgname=hashcash
+version=1.23
+revision=1
+archs="~*-musl"
```
@ericonr I added this line because this package fails to be built with `musl` lib

ericonr replied:
What kind of errors did you get?

g4s8 replied:
I tried to build it in `voidlinux/masterdir-x86_64-musl` Docker the error was:
```
In file included from array.c:5:
array.h:32:5: error: unknown type name 'time_t'
   32 |     time_t expires_before;
      |     ^~~~~~
```

Also, CI on musl also fails.

sgn replied:
Simply add `#include <time.h>` into `array.h` fix the problem.

g4s8 commented:
@ericonr thanks, I've fixed you review comments.

> Also, this is bad code

I'm not the author of this package, I'm just the user. I added it here, because it's missed in Void repositories, and I'm using Void now. FYI it's available in other distro's main repos (e.g. Ubuntu stable has this package). Maybe I can add a patch to remove this code if you think it's critical.

ericonr commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,14 @@
+revision=1
+archs="~*-musl"
+build_style=gnu-makefile
+make_build_args="hashcash docs"
```
```
=> hashcash-1.23_1: running do_install ...
cc -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe         -c -o sha1.o sha1.c
cc sha1.o libsha1.o -o sha1 -Wl,-z,relro -Wl,-z,now -Wl,--as-needed    
```

It's building sha1 during install.

g4s8 replied:
Yes, in `Makefile` it has an option to build either with openssl or with embedded sha1 implementation.

ericonr replied:
What I meant is that this should be moved to `do_build`. And generally speaking I'd expect openssl/libressl SHA1 to perform better.

g4s8 replied:
Sorry, I missed that. Fixed in a new commit: https://github.com/void-linux/void-packages/pull/28379/commits/bad76bf16e5b0a45bfa9ca69a124ad176a35b4f2

g4s8 commented:
@ericonr I added libressl-devel make dependency and openssl runtime dependency in https://github.com/void-linux/void-packages/pull/28379/commits/3b86884358f33411e8b0b714a5e0ef29ce05a85f but it breaks aarch64 and armv7 (in CI checks). Is it better than using embedded sha1 implementation?

ericonr commented:
```
LIBCRYPTO=/usr/lib/libcrypto.a
```

You have to override this value from their makefile. Something like `make_build_args="LIBCRYPTO=$XBPS_CROSS_BASE/usr/lib/libcrypto.a"`

g4s8 commented:
> ```
> LIBCRYPTO=/usr/lib/libcrypto.a
> ```
> 
> You have to override this value from their makefile. Something like `make_build_args="LIBCRYPTO=$XBPS_CROSS_BASE/usr/lib/libcrypto.a"`

@ericonr thanks, that's fixed the problem

sgn commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+revision=1
+archs="~*-musl"
+build_style=gnu-makefile
+make_build_args="LIBCRYPTO=$XBPS_CROSS_BASE/usr/lib/libcrypto.a generic-openssl docs"
```
Sorry if I was missing anything, but what's wrong with `LIBCRYPTO=-lcrypto`?

ericonr replied:
It wants static linking, idk why.

sgn commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+build_style=gnu-makefile
+make_build_args="LIBCRYPTO=$XBPS_CROSS_BASE/usr/lib/libcrypto.a generic-openssl docs"
+makedepends="libressl-devel"
+depends="openssl>=3.0"
```
I don't think we have `openssl` for now.

g4s8 replied:
@sgn If I understand correctly, libressl provides openssl>=3 API, so if we have a dependency on openssl, then both openssl and libressl implementations work.

sgn replied:
OK, libressl provides `openssl-$version`. I think we can just write `libressl` there. Or `openssl`.

ericonr replied:
Does this command call the openssl executable? Otherwise this shouldn't be here. Also you can't depend on the openssl package while it doesn't exist.

g4s8 replied:
@sgn @ericonr OK, I renamed to `libressl>=3.0`

ericonr replied:
No, no version specification. And I don't think this calls the libressl executable, so why is it added as a dependency?

g4s8 replied:
@ericonr according to manual, it should be package name, not executable binary: https://github.com/void-linux/void-packages/blob/master/Manual.md#optional-variables and since I can be sure it's working with libressl 3d version I added >= 3.0. Do you think I should remove version constraint?

ericonr replied:
The only relevant thing in the `libressl` package is the `openssl` executable, equivalent to what an `openssl` package would provide. Since this program doesn't call that executable, why is it a dependency?

abenson replied:
`xbps-src` will correctly detect the right package to add to `depends`, so this is unnecessary.

ericonr replied:
(and the correct package is `libcryptoXX`, not `libressl`.

g4s8 replied:
@abenson @ericonr OK, so I've removed the runtime dependency line, is it good now? edad5fb1071b140a26bbadc2806ccdb95bb467c5

g4s8 commented:
@sgn thanks for review, I've fixed and answered your comments

ericonr commented:
You still haven't added the conflict with outils, and this doesn't need an archs line anymore.

> I'm not the author of this package, I'm just the user.

Yes, but we are still responsible for the quality of upstream software we accept into our repos. And I still don't understand what this package is useful for. That other distros ship a certain package doesn't automatically make it clear a certain bar.

sgn commented on srcpkgs/hashcash/template:
```diff
@@ -0,0 +1,15 @@
+pkgname=hashcash
+version=1.23
+revision=1
+archs="*"
```
Now, it's time to remove it entirely.

g4s8 commented:
@sgn I removed architecture line, @ericonr and added conflict with outils.

> And I still don't understand what this package is useful for

This tool is useful for some network chats or websites, which ask users to solve a hash-task as a proof of work.
